src/ucp/core/ucp_ep.c

```diff
@@ -781,28 +781,31 @@ void ucp_ep_cleanup_lanes(ucp_ep_h ep)
 
     for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
         uct_ep = ep->uct_eps[lane];
-        if (uct_ep != NULL) {
-            ucs_debug("ep %p: purge uct_ep[%d]=%p", ep, lane, uct_ep);
-            uct_ep_pending_purge(uct_ep, ucp_destroyed_ep_pending_purge, ep);
-        }
-    }
-
-    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
-        uct_ep = ep->uct_eps[lane];
-        if (uct_ep == NULL) {
-            continue;
-        }
-
-        ucs_debug("ep %p: destroy uct_ep[%d]=%p", ep, lane, uct_ep);
-        uct_ep_destroy(uct_ep);
-    }
 
-    for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
         /* set UCT EP to failed UCT EP to make sure if UCP EP won't be destoyed
          * due to some UCT EP discarding procedures are in-progress and UCP EP
          * may get some operation completions which could try to dereference
          * its lanes */
         ep->uct_eps[lane] = &ucp_failed_tl_ep;
+
+        if ((uct_ep == NULL) || (uct_ep == &ucp_failed_tl_ep)) {
+            continue;
+        }
+
+        ucs_debug("ep %p: purge uct_ep[%d]=%p", ep, lane, uct_ep);
+        uct_ep_pending_purge(uct_ep, ucp_destroyed_ep_pending_purge, ep);
+
+        /* an attempt to destroy UCT lane can be done several times inside
+         * ucp_ep_cleanup_lanes()/ucp_ep_discard_lanes(), if it is not a case -
+         * UCP EP is going to be destroyed itself (i.e. called from UCP EP
+         * destroy function and UCP EP's ref_cnt has to be '1').
+         * Otherwise, it may happen that UCT EP was destroyed, although some
+         * operations are still in-progress and completions can be received for
+         * this UCT EP - this will lead to undefined behavior */
+        ucs_assert(ep->ref_cnt == 1);
+
+        ucs_debug("ep %p: destroy uct_ep[%d]=%p", ep, lane, uct_ep);
```

Comment:
i'd combine to one trace with line 795, like `purge and destroy` to avoid extra logs flooding

Reply:
fixed